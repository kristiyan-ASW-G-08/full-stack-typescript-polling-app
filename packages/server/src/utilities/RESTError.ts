import ValidationError from '@poll/common/source/types/ValidationError';

const errors = {
  BadRequest: {
    status: 400,
    message: 'Request has wrong format',
  },
  Unauthorized: {
    status: 401,
    message: 'Authentication credentials not valid or missing',
  },
  Forbidden: {
    status: 403,
    message: "You're missing permission to execute this request",
  },
  NotFound: {
    status: 404,
    message: 'Resource not found',
  },
  Conflict: {
    status: 409,
    message: 'Conflict',
  },
  UnprocessableEntity: {
    status: 422,
    message: 'Request cannot be processed',
  },
  TooManyRequests: {
    status: 429,
    message: 'Too many requests',
  },
  InternalServerError: {
    status: 500,
    message: 'Internal server error',
  },
};

export default class RESTError extends Error {
  public status: number;

  public message: string;

  public data?: ValidationError[] | string;

  public constructor(
    status: number,
    message: string,
    data?: ValidationError[] | string,
  ) {
    super();
    Object.setPrototypeOf(this, RESTError.prototype);
    this.status = status;
    this.message = message;
    this.data = data;
  }
}

export { RESTError, errors };
