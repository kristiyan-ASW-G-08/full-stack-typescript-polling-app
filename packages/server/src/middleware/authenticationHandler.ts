import { verify } from 'jsonwebtoken';
import { Request, Response, NextFunction } from 'express';
import RESTError, { errors } from '@utilities/RESTError';

const authenticationHandler = (
  req: Request,
  res: Response,
  next: NextFunction,
): void => {
  const { JWT_SECRET } = process.env;
  const authHeader = req.get('Authorization');
  const { status, message } = errors.Unauthorized;
  const error = new RESTError(status, message);
  if (!authHeader) {
    throw error;
  }
  const token = authHeader.split(' ')[1];
  // @ts-ignore
  const { userId } = verify(token, JWT_SECRET);
  if (!userId) {
    throw error;
  }
  req.userId = userId;
  next();
};
export default authenticationHandler;
