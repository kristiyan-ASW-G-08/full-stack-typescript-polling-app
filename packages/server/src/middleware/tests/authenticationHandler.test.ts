import mongoose from 'mongoose';
import httpMocks from 'node-mocks-http';
import jwt from 'jsonwebtoken';
import authenticationHandler from '@customMiddleware/authenticationHandler';
import RESTError from '@utilities/RESTError';
import ValidationError from '@poll/common/source/types/ValidationError';

jest.mock('@utilities/RESTError');

const RESTErrorMock = RESTError as jest.MockedClass<typeof RESTError>;

jest.spyOn(jwt, 'verify');
RESTErrorMock.mockImplementationOnce(
  (status: number, message: string, data?: ValidationError[] | string) => ({
    status,
    message,
    data,
    name: 'error',
  }),
);

describe('authenticationHandler', (): void => {
  const { JWT_SECRET } = process.env;
  afterEach(() => jest.clearAllMocks());
  afterAll(() => jest.restoreAllMocks());
  it(`should add userId to req when the authorization header is valid`, async (): Promise<
    void
  > => {
    expect.assertions(5);
    const nextMock = jest.fn();
    const userId = mongoose.Types.ObjectId();
    const token = jwt.sign(
      {
        userId,
      },
      JWT_SECRET,
      { expiresIn: '1h' },
    );
    const reqMock = httpMocks.createRequest({
      method: 'POST',
      url: '/',
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    const resMock = httpMocks.createResponse();
    authenticationHandler(reqMock, resMock, nextMock);

    expect(jwt.verify).toHaveBeenCalledTimes(1);
    expect(jwt.verify).toHaveBeenCalledWith(token, JWT_SECRET);
    expect(jwt.verify).toHaveReturnedTimes(1);
    expect(reqMock.userId).toMatch(userId.toString());
    expect(nextMock).toBeCalledTimes(1);
  });
  it('should throw an error with a status of 401: Unauthorized when there is no authorization header', async (): Promise<
    void
  > => {
    expect.assertions(1);
    const nextMock = jest.fn();
    const reqMock = httpMocks.createRequest({
      method: 'POST',
      url: '/',
    });
    const resMock = httpMocks.createResponse();

    expect((): void =>
      authenticationHandler(reqMock, resMock, nextMock),
    ).toThrowErrorMatchingSnapshot();
  });
  it("should throw an error with a status of 401: Unauthorized when the decoded jwt token doesn't contain a user id ", async (): Promise<
    void
  > => {
    expect.assertions(1);
    const nextMock = jest.fn();
    const token = jwt.sign({}, JWT_SECRET, { expiresIn: '1h' });
    const reqMock = httpMocks.createRequest({
      method: 'POST',
      url: '/',
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    const resMock = httpMocks.createResponse();

    expect((): void =>
      authenticationHandler(reqMock, resMock, nextMock),
    ).toThrowErrorMatchingSnapshot();
  });
});
