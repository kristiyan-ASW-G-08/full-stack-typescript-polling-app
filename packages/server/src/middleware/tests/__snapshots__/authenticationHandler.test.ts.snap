// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticationHandler should throw an error with a status of 401: Unauthorized when the decoded jwt token doesn't contain a user id  1`] = `""`;

exports[`authenticationHandler should throw an error with a status of 401: Unauthorized when there is no authorization header 1`] = `""`;
